components/family-studies/migrations/src/main/java/org/phenotips/studies/family/migrations/R71507PhenoTips3423DataMigration.java

```diff
@@ -0,0 +1,558 @@
+                session.clear();
+                ((XWikiHibernateStore) getStore()).saveXWikiDoc(xDocument, context, false);
+                session.flush();
+            } catch (DataMigrationException e) {
+                this.logger.error("Error when saving XWiki document {}: [{}]", docName, e.getMessage());
+            }
+
+        }
+        return null;
+    }
+
+    // used to store data about a (potentially multi-generation) path (in a graph sense)
+    // from partners to their relationships
+    private static class PartnerPath
+    {
+        private Integer parentID;
+        private List<Integer> path;
+
+        PartnerPath(Integer parentID)
+        {
+            this.parentID = parentID;
+            this.path = new LinkedList<>();
+        }
+
+        public void addPathElement(Integer virtualNodeID)
+        {
+            this.path.add(virtualNodeID);
+        }
+
+        public Integer getParentID()
+        {
+            return this.parentID;
+        }
+
+        public List<Integer> getPath()
+        {
+            return this.path;
+        }
+
+        public String toString()
+        {
+            if (this.path.size() == 0) {
+                return this.parentID.toString() + " (direct)";
+            } else {
+                return this.parentID.toString() + " -> <" + this.path.toString() + ">";
+            }
+        }
+    }
+
+    private boolean updatePedigree(BaseObject pedigreeXObject, XWikiContext context, String documentName)
+    {
+        String oldPedigreeAsText = pedigreeXObject.getStringValue(PEDIGREECLASS_JSONDATA_KEY);
+        if (!StringUtils.isEmpty(oldPedigreeAsText)) {
+            String convertedPedigree = this.convertPedigreeData(oldPedigreeAsText);
+            pedigreeXObject.set(PEDIGREECLASS_JSONDATA_KEY, convertedPedigree, context);
+            return true;
+        }
+        return false;
+    }
+
+    private String convertPedigreeData(String oldFormatData)
+    {
+        JSONObject oldPedigreeData = new JSONObject(oldFormatData);
+
+        this.logger.debug("Old pedigree: [{}]", oldPedigreeData.toString());
+
+        // create  asceleton JSON with all data structures present but blank
+        JSONObject newPedigree = this.createBlankNewPedigreeJSON();
+
+        newPedigree.put(PEDIGREE_PROBAND_KEY, oldPedigreeData.optInt(OLD_PEDIGREE_PROBAND_KEY, 0));
+
+        JSONObject pedigreeSettings = oldPedigreeData.optJSONObject(OLD_PEDIGREE_SETTINGS_KEY);
+        if (pedigreeSettings != null) {
+            newPedigree.put(PEDIGREE_SETTINGS_KEY, pedigreeSettings);
+        }
+
+        // Note1: the assumption is that pedigree is consistent and correct, if a required key
+        //        is missing an exception will be thrown (e.g. by getJSONArray()), which is expected
+        //
+        // Note2: migrated pedigree will preserve all node IDs, so no ID mapping is required
+
+        JSONArray oldData = oldPedigreeData.getJSONArray(OLD_PEDIGREE_GRAPH_KEY);
```

Comment:
why not `optJSONArray()`?

Reply:
because if there is no array we want to throw an exception (and abort conversion for this family), this is an abnormal situation that we can't handle